src/Api/NavigationUtils.ts

```diff
@@ -150,9 +174,218 @@ export const getResolveConditionsResult = (
+}): {
+  commonStaticNavigator: StaticTreeNavigator,
+  targetPathFromCommonNavigator: string[],
+  sourcePathFromCommonNavigator: string[],
+} => {
+  let leftDepth = currentStaticTreeScreen.depth ?? 0
+  let rightDepth = finalStaticTreeScreen.depth ?? 0
+  let leftParent: StaticTreeNode = currentStaticTreeScreen
+  let rightParent: StaticTreeNode = finalStaticTreeScreen
+  const leftPath: string[] = []
+  const rightPath: string[] = []
+  const biggerDepth = Math.max(leftDepth, rightDepth)
+  for (let i = biggerDepth; i > 1; i--) {
+    if (leftDepth === i) {
+      leftParent = is(leftParent.getParent())
+      leftPath.unshift(is(leftParent.routeName))
+      leftDepth--
+    }
+    if (rightDepth === i) {
+      rightParent = is(rightParent.getParent())
+      rightPath.unshift(rightParent.routeName)
+      rightDepth--
+    }
+    if (leftParent?.routeName === rightParent?.routeName) {
+      break
+    }
+  }
+
+  return {
+    commonStaticNavigator: leftParent as StaticTreeNavigator,
+    targetPathFromCommonNavigator: rightPath,
+    sourcePathFromCommonNavigator: leftPath,
+  }
+}
+
+export const transformForNearestExistingNavigator = (
+  action: NavigationAction,
+  getRootState: () => NavigationState,
+  tree: StaticTreeNavigator,
+): NavigationAction => {
+  if (action.isTransformed ?? false) {
+    return action
+  }
+
+  const navigationState = getRootState()
+  const activeRouteName = getActiveLeafRoute(navigationState).name
+  const targetRouteName = last(getRoutePathFromAction(action) ?? [])
+
+  const currentStaticTreeScreen = isNotEmptyString(activeRouteName)
+    ? findStaticTreeScreen(tree, activeRouteName)
+    : undefined
+  const finalStaticTreeScreen = findStaticTreeScreen(tree, targetRouteName)
+  if (currentStaticTreeScreen == null || finalStaticTreeScreen == null) {
+    throw Error(`Missing static tree screen for route name: ${currentStaticTreeScreen == null ? activeRouteName : targetRouteName}`)
+  }
+  if (currentStaticTreeScreen === finalStaticTreeScreen) {
+    return action
+  }
+
+  const {
+    commonStaticNavigator,
+    sourcePathFromCommonNavigator,
+    targetPathFromCommonNavigator,
+  } = getCommonStaticNavigatorWithPaths({
+    currentStaticTreeScreen,
+    finalStaticTreeScreen,
+  })
+
+  if (arePathsStartingEqually(sourcePathFromCommonNavigator, targetPathFromCommonNavigator)) {
+    return action
+  }
+
+  let nextAction
+  if (action.type === 'NAVIGATE') {
+    nextAction = createNavigateActionForPath(targetPathFromCommonNavigator, action)
+  } else {
+    nextAction = action
+  }
+
+  log.debug('TRANSFORMED ACTION', {
+    originalAction: action,
+    nextAction,
+  })
+
+  return {
+    ...nextAction,
+    navigatorId: commonStaticNavigator?.id,
+    isTransformed: true,
+  }
+}
+
+export const calculateStaticTreeDepth = (tree: StaticTreeNodeDeclaration, parent?: StaticTreeNavigator, depth = 0): StaticTreeNode => {
+  const treeWithDepth: StaticTreeNode = {
+    ...tree as StaticTreeNode,
```

Comment:
Because `tree.screens` are still `StaticTreeNodeDeclaration` but `treeWithDepth.screens` should already be `StaticTreeNode` and that only happens in the `screens.map`. I could maybe spread `screens` separately and cast only that?